src/Silverback.Integration/Diagnostics/DiagnosticsConsumer.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Silverback.Messaging;
+using Silverback.Messaging.Broker;
+using Silverback.Messaging.Messages;
+
+namespace Silverback.Diagnostics
+{
+    public abstract class DiagnosticsConsumer : Consumer

```

Comment:
It works properly with the intermediate base class `DiagnosticsConsumer` and `DiagnosticsProducer` but I don't know if I like it. I see this as a side aspect and I would have just used an helper from within the regular `Consumer`.

Also, when other stuff get added, it will be confusing for whoever implements a extra consumer or producer to choose which is the proper class to extend and doesn't give the freedom to choose exactly which aspect to inherit.

I think we should move this logic into something pluggable...but it can be done later on.

What I imagine (completely made up right now):

```c#
public abstract class Consumer : ...
{
    public readonly IEnumerable<IConsumerBehavior> _behaviors;
    public Consumer(IEnumerable<IConsumerBehavior> behaviors) => _behaviors = behaviors;
    
        protected Task HandleMessage(byte[] message, IEnumerable<MessageHeader> headers, IOffset offset)
        {
            _behaviors.ForEach(b => b.Consuming(message, headers, offset));

            try
            {
                //...actual consume logic...

                _behaviors.ForEach(b => b.Consumed(message, headers, offset));
            }
            catch (Exception ex)
            {
                _behaviors.ForEach(b => b.Failed(message, headers, offset));
            }
        }
}
```

So that trancing, logging, activity and all other aspects can be moved into an implementation of `IConsumerBehavior` and easily activated/deactivated in the startup.

The same can of course easily be adapted into an `IProducerBehavior` (with the difference that it can also modify the actual message being sent).

Anyhow, I suggest to make it work like this and refactor it later. If you don't mind, I can provide the basic implementation/prototype for these "behaviors" in a branch, maybe tomorrow while traveling to Bern.

Reply:
Note: ...discussion is going on, this is not up-to-date anymore

Reply:
The `IConsumerBehavior` and `IProducerBehavior` could be implemented following the same logic of the `IBehavior`.

https://github.com/BEagle1984/silverback/blob/75078c8aec459bae32a027fc069eef8099db771c/src/Silverback.Core/Messaging/Publishing/Publisher.cs#L85

https://github.com/BEagle1984/silverback/blob/75078c8aec459bae32a027fc069eef8099db771c/src/Silverback.Core/Messaging/Publishing/IBehavior.cs#L13